### PR TITLE
Create CommonCrawler.py

### PR DESCRIPTION
**Pull request by maxachis**
#### Fixes

Issue #40 

#### Description

Creates the Common Crawler module, a module for interfacing with the Common Crawl dataset and extracting urls. Full details can be found within the README file of the provided module.

#### Testing

Unit and integration tests are included in the `tests` directory and can be run according to instructions provided in the README file included.

### CODE REVIEW

**Comment by josh-chamberlain** [comment]
@maxachis I hope it's OK I pulled the content from your comment on the issue into the PR description. the "fixes #" at the start links it to the other one. You might want to delete the issue comment.

One more style thing (that I didn't warn you about) is a request that you use the [Issue](https://github.com/Police-Data-Accessibility-Project/docs/blob/main/meta/operations/staff/github-issue-template.md) and [Pull Request](https://github.com/Police-Data-Accessibility-Project/docs/blob/main/meta/operations/staff/github-pull-request-template.md) templates.

**Review by josh-chamberlain — DISMISSED** [review]
Could you please rename this to `common_crawler` so it fits the paradigm of...most of the other ones, but not all of them? Also, if you add a requirements.txt file, and steps to test it, that would be great. I was able to reproduce this by creating a venv and installing the requirements manually.

**Comment by josh-chamberlain** [comment]
Aside from my comments, this makes sense to me—over to @mbodeantor 

Also, for consideration:

- how can we easily avoid, every time someone runs this, grabbing the first page and working with the same URLs over and over? even something dumb like randomizing the page number could work
- it would be interesting to run this against all the agency homepages currently in the database, like `https://city.com/police/*`. That overlaps with the functionality of the [sitemap scraper](https://github.com/Police-Data-Accessibility-Project/scrapers/tree/main/utils/sitemap_scraper), but might work better (or worse)

**Comment by maxachis** [comment]
> * how can we easily avoid, every time someone runs this, grabbing the first page and working with the same URLs over and over? even something dumb like randomizing the page number could work

@josh-chamberlain A simple way to handle this would be having a small cache or database indicating what pages from what index have been pulled. A small SQLite database for more discretely tracking both pages and which index is used, that is pushed with the repo, could work. One could go even simpler and have a small JSON file which keeps track of the last page done for each index (and assumes that the next page pulled will always be one greater than the last).

**Comment by maxachis** [comment]
@josh-chamberlain I have updated the Pull Request and the module according to your request, and made additional enhancements. To wit:

1. The Pull Request has been reformatted to the Pull Request Template
2. The module has been updated such that searches are cached via a file contained within the module itself, ensuring that the same pages are not pulled repeatedly for the same parameters. 
3. The module has also been updated such that all results are written to a csv file also contained within the module itself. Currently this contains 200 urls. 
4. I modified the module to be more, well, modular, decoupled, and testable.
5. Similarly, I enhanced the unit and integration test suite. These can also be instructive in demonstrating the intended functionality of the script. 

I'm ready for your review at your convenience.

**Comment by josh-chamberlain** [comment]
thanks @maxachis ! I think you reference a README file, and I remember seeing it, but I don't see it now.

**Comment by maxachis** [comment]
> thanks @maxachis ! I think you reference a README file, and I remember seeing it, but I don't see it now.

@josh-chamberlain  Ah, I did the ol' "commit but didn't push" mistake. That's fixed now.

**Comment by mbodeantor** [comment]
Looks like an extra requirements.txt slipped in here.

**Comment by mbodeantor** [comment]
Can you also simplify the code a bit? OOP is fine but there is a lot of extra abstraction in here that makes it hard to follow. Like storing the page count for the crawl loop on the cache object

**Comment by mbodeantor** [comment]
@maxachis Thanks for those changes. While I appreciate the idea of extensibility, I still would prefer a reduction in the number of classes. As an organization with developers coming and going, we really have to lean towards readability rather than future proofing.

**Comment by maxachis** [comment]
@mbodeantor Understood. My latest changes have consolidated the cache and crawler logic. To wit:

1. Cache logic is now a single class with substantially simplified logic. 
2. Crawler logic is now also a single class (albeit with a dataclass and named tuple included to better organize results), and cache logic has been migrated out of the crawler logic such that it now occurs in the scope of main, rather than being embedded within the crawler manager. 
3. I additionally removed unnecessary methods and unit tests. 

Each file, absent the utils file and argparser file, is now defined by a single primary class, occasionally with small supporting dataclasses or named tuples to organize method results. Contrary to before, no class includes a subordinate class. All files remain separated into clear logical domains, and main now clearly demarcates when a particular domain is being accessed.

**Comment by mbodeantor** [comment]
@maxachis Awesome, thanks for making those changes. The only thing tripping me up still are the tests, when I tried to run them using the instructions in the readme I got an error "fixture 'mocker' not found"

**Comment by maxachis** [comment]
@mbodeantor Ah, the dreaded 'works on my machine', dilemma. As best I can tell, it's becuase I didn't include pytest-mock in the requirements.txt. That's been updated. Try cloning the latest version of the repo, installing everything in requirements.txt, and repeating the command.

**Review by mbodeantor — APPROVED** [review]
